File: Febrero2016/haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g 
    | isEmpty g = True
    | length (vertices g) == 1 = True
    | length (vertices g) == 2 = False
    | otherwise = isEulerian' (vertices g)
        where
            isEulerian' [] = True
            isEulerian' (x:xs)
                | even (degree g x) = isEulerian' xs
                | otherwise = False
-- H.2)
{-remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = remove' (vertices g') g' 
    where
        g' = deleteEdge g (v,u)
        remove' (x:xs) g'
            | (degree g' x) == 0 = remove' xs (deleteVertex g' x)
            | otherwise = remove' xs g'-}


remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u)
    | (degree g' v) == 0 && (degree g' u) /= 0 = (deleteVertex g' v)
    | (degree g' u) == 0 && (degree g' v) /= 0 = (deleteVertex g' u)
    | (degree g' u) == 0 && (degree g' v) == 0 = deleteVertex (deleteVertex g' u) v
    | otherwise = g'
    where
        g' = deleteEdge g (v,u)

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = aux [v0] v0 (head(successors g v0)) g
    where
        aux ciclo v u g
            | v0 == u = (g',ciclo ++ [v0])
            | otherwise = aux (ciclo ++ [u]) u (head(successors g' u)) (g')
                where
                    g' = remove g (v,u) 

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys = ys
connectCycles (x:xs) (y:ys)
    | x == y = [x] ++ ys ++ xs
    | otherwise = [x] ++ connectCycles xs (y:ys)

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = (head[x | x <- (vertices g), elem x cycle])

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g
    | not(isEulerian g) = error "el grafo no es euleriano"
    | isEmpty g = [] 
    | length (vertices g) == 1 = [head(vertices g)]
    | otherwise = aux g' p
        where
            (g',p) = extractCycle g (head(vertices g))
            aux g p
                | isEmpty g = p
                | otherwise = aux (fst grafonuevo) (connectCycles p (snd grafonuevo))
                    where
                        verticecomun = vertexInCommon g p
                        grafonuevo = (extractCycle g verticecomun)







data Vertex = A | B | C | D | E | F | G | H | I | J deriving (Show,Eq,Enum,Ord)

g0n :: Graph Vertex -- not eulerian
g0n = mkGraphEdges [] []

g1 :: Graph Vertex -- eulerian
g1 = mkGraphEdges [A] []

g2n :: Graph Vertex -- not eulerian
g2n = mkGraphSuc [A,B] suc
    where
      suc A = [B]
      suc B = [A]

g3 :: Graph Vertex -- eulerian
g3 = mkGraphEdges [A .. C]
                  [(A,B), (B,C), (C,A)]

g5 :: Graph Vertex
g5 = mkGraphEdges [A .. E]
                  [(A, C), (A, D), (B, C), (B, E), (C, D), (C, E)]

g6 :: Graph Vertex
g6 = mkGraphSuc vertices suc -- eulerian
    where
      vertices = [A .. F]
      suc A = [B,E]
      suc B = [A,C,D,E]
      suc C = [B,D,E,F]
      suc D = [B,C,E,F]
      suc E = [A,B,C,D]
      suc F = [C,D]

{-
  A--B--D--F
     |  |
     C--E
-}
g6n  = mkGraphEdges vertices edges -- not eulerian
    where
      vertices =  [A .. F]
      edges = [(A,B),(B,C),(B,D),(D,E),(D,F),(C,E)]

{-
  A--B--D--F
   \ |  |
     C--E--G
-}
g7n  = mkGraphSuc vertices suc -- not eulerian
    where
      vertices = [A .. G]
      suc A = [B,C]
      suc B = [A,C,D]
      suc C = [A,B,E]
      suc D = [B,F,E]
      suc E = [C,D,G]
      suc F = [D]
      suc G = [E]

g10 = mkGraphSuc vertices suc -- eulerian
    where
      vertices = [A .. J]
      suc A = [E,J,I,B]
      suc B = [A,I,H,C]
      suc C = [D,G,H,B]
      suc D = [E,F,G,C]
      suc E = [D,F,J,A]
      suc F = [E,D,G,J]
      suc G = [D,C,H,F]
      suc H = [G,I,C,B]
      suc I = [A,B,J,H]
      suc J = [E,F,A,I]

File: Febrero2016/eulerianCycle/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 *
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        if(g.isEmpty() || g.vertices().size() == 1){
            return true;
        }else if (g.vertices().size() == 2){
            return false;
        }else{
            for(V v:g.vertices()){
                if(g.degree(v) % 2 != 0){
                    return false;
                }
            }
        }
        return true;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v,u);
        for(V ver :g.vertices()){
            if(g.degree(ver) == 0){
                g.deleteVertex(ver);
            }
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> ciclo = new ArrayList<>();
        ciclo.append(v0);
        V v = v0;

        while (!g.successors(v).isEmpty()){
            V u = g.successors(v).iterator().next();
            ciclo.append(u);
            remove(g,v,u);
            v = u;
        }
        return ciclo;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
    	List<V> sol = new ArrayList<>();
        if(xs.isEmpty()){
            for(V y:ys) xs.append(y);
        }else{
            for (V v : xs){
                if(v == ys.get(0)){
                    for(V y:ys) {
                        sol.append(y);
                    }
                }else{
                    sol.append(v);
                }
            }
            while (!xs.isEmpty()){
                xs.remove(0);// Limpiar xs
            }
            for (V v : sol) {
                xs.append(v);  // Rellenar xs con los elementos de sol
            }
        }
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
    		for(V v:cycle){
                if(g.vertices().isElem(v)){
                    return v;
                }
            }
    		return null;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
    	if(!isEulerian(g)){
            return null;
        }else{
            List<V> cicloparcial = extractCycle(g,g.vertices().iterator().next());
            while(!g.isEmpty()){
                connectCycles(cicloparcial,(extractCycle(g,vertexInCommon(g,cicloparcial))));
            }
            return cicloparcial;
        }
    }
}
